kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyBackendHandler.java

```diff
@@ -5,99 +5,225 @@
+                pendingServerFlushes = false;
+            }
         }
+        LOGGER.trace("/READ");
     }
 
-    @Override
-    public void channelInactive(ChannelHandlerContext ctx) {
-        frontendHandler.closeNoResponse(inboundCtx.channel());
+    /**
+     * Called by the {@link ProxyChannelStateMachine} when the bach from the downstream/client side is complete.
+     */
+    public void flushToServer() {
+        if (serverCtx != null) {
+            final Channel serverChannel = serverCtx.channel();
+            if (pendingServerFlushes) {
+                pendingServerFlushes = false;
+                serverChannel.flush();
+            }
+            if (!serverChannel.isWritable()) {
+                proxyChannelStateMachine.onServerUnwritable();
+            }
+        }
     }
 
-    @Override
-    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
-        // If the frontEnd has an exception handler for this exception
-        // it's likely to have already dealt with it.
-        // So only act here if its un-expected by the front end.
-        if (exceptionMapper.mapException(cause).isEmpty()) {
-            LOGGER.atWarn()
-                    .setCause(LOGGER.isDebugEnabled() ? cause : null)
-                    .addArgument(cause != null ? cause.getMessage() : "")
-                    .log("Netty caught exception from the backend: {}. Increase log level to DEBUG for stacktrace");
-            frontendHandler.closeNoResponse(ctx.channel());
+    /**
+     * Callback from the {@link ProxyChannelStateMachine} triggered when it wants to apply backpressure to the <em>upstream/server</em> connection
+     */
+    public void blockServerReads() {
+        if (serverCtx != null) {
+            serverCtx.channel().config().setAutoRead(false);
+        }
+    }
+
+    /**
+     * Callback from the {@link ProxyChannelStateMachine} triggered when it wants to remove backpressure from the <em>upstream/server</em> connection
+     */
+    public void unblockServerReads() {
+        if (serverCtx != null) {
+            serverCtx.channel().config().setAutoRead(true);
+        }
+    }
+
+    /**
+     * Called by the {@link ProxyChannelStateMachine} on entry to the {@link ProxyChannelState.Closed} state.
+     */
+    public void inClosed() {
+        if (serverCtx != null) {
+            Channel outboundChannel = serverCtx.channel();
+            if (outboundChannel.isActive()) {
+                outboundChannel.writeAndFlush(Unpooled.EMPTY_BUFFER).addListener(ChannelFutureListener.CLOSE);
+            }
         }
     }
+
+    @Override
+    public String toString() {
+        // Don't include proxyChannelStateMachine's toString here
+        // because proxyChannelStateMachine's toString will include the backends toString
+        // and we don't want a SOE.
+        return "KafkaProxyBackendHandler{" + ", serverCtx=" + serverCtx + ", proxyChannelState=" + this.proxyChannelStateMachine.currentState()
```

Comment:
Have you checked that the logging still has reasonable detail? I can _imagine_ it does.

Reply:
No, I haven't watched logs in anger. I probably should...